data/scripts/actions/spellbook.lua

```diff
@@ -0,0 +1,35 @@
+local spellbook = Action()
+
+function spellbook.onUse(player, item, fromPosition, target, toPosition, isHotkey)
+	local text = ""
+	local spells = {}
+	for _, spell in ipairs(player:getInstantSpells()) do
+		if spell.level ~= 0 then
+			if spell.manapercent > 0 then
+				spell.mana = spell.manapercent .. "%"
+			end
+			spells[#spells + 1] = spell
+		end
+	end
+
+	table.sort(spells, function(a, b) return a.level < b.level end)
+
+	local prevLevel = -1
+	for i, spell in ipairs(spells) do
+		local line = ""
+		if prevLevel ~= spell.level then
+			if i ~= 1 then
+				line = "\n"
+			end
+			line = line .. "Spells for Level " .. spell.level .. "\n"
+			prevLevel = spell.level
+		end
+		text = text .. line .. "  " .. spell.words .. " - " .. spell.name .. " : " .. spell.mana .. "\n"
```

Comment:
You might wanna store the strings into a table and then call, `table.concat(t)`, because this currently creates a lot of temporary strings.

Reply:
This was actually the original spellbook lua which is implemented at the moment, just moved it to scripts to show how the lua file setup would look like.
However fixed what you suggested in [9c085d8](https://github.com/otland/forgottenserver/pull/2558/commits/9c085d844c9bc73bed7f6056131be74ed4a6ce03)

Reply:
I think you can also apply the same technique when creating a line, but looks good anyway.

Reply:
I guess you mean something like this?
```lua
for i, spell in ipairs(spells) do
	if prevLevel ~= spell.level then
		if i == 1 then
			text[#text == nil and 1 or #text+1] = "Spells for Level "
		else
			text[#text+1] = "\nSpells for Level "
		end
		text[#text+1] = spell.level
		text[#text+1] = "\n"
		prevLevel = spell.level
	end
	text[#text+1] = spell.words
	text[#text+1] = " - "
	text[#text+1] = spell.name
	text[#text+1] = " : "
	text[#text+1] = spell.mana
	text[#text+1] = "\n"
end
```

Reply:
Looks terrible :P, i wonder if this can be abstracted somehow, my lua-fu isn't good enough to tell you.

Reply:
could be done with another loop inside parsing 'spell' actually and then glue it together but then again it's kinda bad to setup and I'd rather not put another loop in for basicly no 'real' reason

Reply:
Use my spellbook script instead, it's best optimized and the closest to the global one.
Tables requires twice the access number so it's not really a good approach. In this particular case the best option would be to parse all spells on startup and create stored versions for each class divided by level and magic level, this way you could easily retrieve cached versions without performing this whole loop every damn time. But since I'm too lazy to do this on startup and store this table in a lib file and reload it, use this "not the best but still very good" approach below:

`function onUse(player, item, fromPosition, target, toPosition, isHotkey)
	local text = ""
	local tlvl = {}
	local tml = {}

	for _, spell in ipairs(player:getInstantSpells()) do
		if spell.level ~= 0 or spell.mlevel ~= 0 then
			if spell.manapercent > 0 then
				spell.mana = spell.manapercent .. "%"
			end
			if spell.level > 0 then
				tlvl[#tlvl+1] = spell
			elseif spell.mlevel > 0 then
				tml[#tml+1] = spell
			end
		end
	end

	table.sort(tlvl, function(a, b) return a.level < b.level end)
	local prevLevel = -1
	for i, spell in ipairs(tlvl) do
		local line = ""
		if prevLevel ~= spell.level then
			if i ~= 1 then
				line = "\n"
			end
			line = line .. "Spells for Level " .. spell.level .. "\n"
			prevLevel = spell.level
		end
		text = text .. line .. "  " .. spell.words .. " - " .. spell.name .. " : " .. spell.mana .. "\n"
	end
	text = text.."\n"
	table.sort(tml, function(a, b) return a.mlevel < b.mlevel end)
	local prevmLevel = -1
	for i, spell in ipairs(tml) do
		local line = ""
		if prevLevel ~= spell.mlevel then
			if i ~= 1 then
				line = "\n"
			end
			line = line .. "Spells for Magic Level " .. spell.mlevel .. "\n"
			prevmLevel = spell.mlevel
		end
		text = text .. line .. "  " .. spell.words .. " - " .. spell.name .. " : " .. spell.mana .. "\n"
	end


	player:showTextDialog(item:getId(), text)
	return true
end
`